TEP: Contract source registry (#91)
ton-blockchain/TEPs: text/0092-contract-source-registry.md

Reviewer: Ok

Reviewer 2: UQCWb4npmDTu3tkvQVHCDBpvDYSMLgyPV2Cvfxhpy3ucRHxo

Reviewer 2: How to earn ton and other coins please help any contract swap link

Reviewer 2: How to start

Reviewer 3: Help me please,my wallet
 UQCO6a6PAQItRzxVP707Z_Gq76fDGWSwQOE_V3BRDi48zoW_

Reviewer 4: UQD72Yp47y8338TpQgnUrJy53XH_qsZsybpscA01HJAN_sDQ

Reviewer 5: text/0092-contract-source-registry.md

Reviewer 6: d916add3dadefb1c131c90f8a1c7a4ad45330eb5

Reviewer 7: 010203

Since TON follows the general principle that "code is law", users who participate in a smart contract, by sending it transactions or depositing funds, effectively agree to the terms coded in this smart contract. Just as nobody would expect users to sign a legal agreement written in a language that they cannot read, the on-chain bitcode is not enough to reassure participants of the terms of agreement.

TON smart contracts originate from higher level languages such as FunC which are human-readable and have been compiled. Since these source files are human-readable, offering a well-known registry where they can be located benefits the community. Making sure that this infrastructure is decentralized would guarantee that this critical resource is always available to anyone under equal access.

# Guide

The standard covers the interactions between multiple relevant stakeholders in the community:

* **Contract developers** - Any developer who writes contract source code, compiles it and deploys on-chain. These developers do not necessarily bother with publishing or paying for the source code to be verified.

> * **Source-code uploaders** - Any community member that has access to source code that matches an on-chain code cell and is interested in publishing this source code and paying for its verification. This is not necessarily the devloper of the contract.

Reviewer 8: ```suggestion
* **Source-code uploaders** - Any community member that has access to source code that matches an on-chain code cell and is interested in publishing this source code and paying for its verification. This is not necessarily the developer of the contract.
```

Reviewer 9: **__**

* **Source-code verifiers** - A permissionless group of community validators that are performing a validation process to check that a specific source-code yields the required compiled on-chain result and publish signed attestations towards this.

> * **Source-code displayers** - Normally explorers (see https://ton.app/explorers) that will display the verified source-code to their users by incorporating a widget in their websites.

Reviewer 4: UQD72Yp47y8338TpQgnUrJy53XH_qsZsybpscA01HJAN_sDQ

* If a smart contract supports code updates, once its code is changed, the new code hash will no longer be found in the registry until a source code matching the new code hash is uploaded.

* IPFS is used as a **temporary** form of storage only until TON Storage is released. All participants commit to performing a full migration of all verified sources to TON Storage when possible.

* A single participating **verifier** runs a backend that exposes HTTP API that receives a list of source files, compiles them and returns a signed proof over the resulting code hash. The public key for this verifier is registered in the verifier registry contract.

* To encourage redundancy of verifiers, each verifier is encouraged to run multiple backend instances, each with its own public key. Verification then requires a quorum of backends to work and not all of them, for example 3 out of 5. This is optional and a verifier can choose to run a single backend with a single public key too. This redundancy allows for easy participation of decentralized oracle networks like Orbs Network in the protocol (networks that have dozens of independent validators) as a single verifier.

## sources.json

> JSON file provided by a specific **verifier** for a specific **contract code hash** containing the URLs of source-code files and verification attestations. Fields of this file include:

Reviewer 8: It should also include the original signatures (which are done on the `data` part of the json)
```
{
   data: {
      codeHash:...
  },
  sigs: [{sig: '...', pubKey: '...'}]
}
```

otherwise, the UI will not be able to display the signatures used to verify the contract.

Author: I know that it's needed, but then it's a bit circular. Because the signature is over code hash + sources_json_url and the sources_json_url is derived from a hash over its data and the signature is in the data. Do you have any simple way to overcome this?

Reviewer 8: What if signatures are stored on chain within the source item contract? 
Each signature is 512bit, so this would add another 2-3 cells (assuming 5 verifier multi sig threshold)

This has interesting implications in case we retire and replace one or more of the original verifying nodes, so it would be good to decide how clients should handle such case (i.e. that the public key used to sign can no longer be found in the verifier registry).

Reviewer 10: I have with my locked crypto

* `codeHash` - SHA256 of the code cell CellRepr (https://ton-blockchain.github.io/docs/tvm.pdf 3.1.5), e.g. E/XXoxbG124QU+iKxZtd5loHKjiEUTcdxcW+y7oT9Q4=
* `sources` - Array of sources files, each with the following fields:
    * `url` - Human-readable source-code URL on IPFS and later TON Storage, e.g. [ipfs://QmWQE1HhYuWieFcocfZZvK4gvJqgswNpj3zUCvdPNCC12T](https://tonsource.infura-ipfs.io/ipfs/QmWQE1HhYuWieFcocfZZvK4gvJqgswNpj3zUCvdPNCC12T)
    * `originalFilename` - e.g. wallet-v3.fc or imports/stdlib.fc
* `compilerType` - Compiler used to compile the verified sources, e.g. `func`

> * `compileVersion` - Version of compiler, e.g. 0.2.0

Reviewer 8: ```suggestion
* `compilerVersion` - Version of compiler, e.g. 0.2.0
```

* `commandLine` - Command-line used to compile sources, e.g. func -o output.fif -SPA wallet-v3.fc stdlib.fc

## Verifier registry contract

A smart contract deployed to TON mainnet that holds a mapping between a **verifier id** to the **verifier details** which include the list of backends, their public keys and quorum configuration. To prevent spam in this registry, we propose that each verifier will deposit in the contract a sum of 1,000-10,000 TON coin. This sum will be returned when the verifier unregisters.

### Actions

> * `update_verifier(verifier_id, backend_endpoints, quorum_config)` - If the verifier does not exist, ensures it deposits the required amount and adds to the registry. Otherwise updates details in the registry. The address that sends this update message is stored in the registry as the admin address and only it can update. The quorum config contains the list of public keys and how many are needed for quorum.

Reviewer 8: There probably should be a public key sent with this op as well, so that future operations (update endpoints, remove verifier) for an existing verifier can be authorized.

Author: I think the simplest behavior is that the address that sends the update message is regarded as the admin address and only this address can update later. So if the admin is a wallet contract, it would send the internal message of the update and on the first update (insert) it would be set.

Reviewer 11: Yes In general I think verifying the sender is more elegant ( and more gas and storage efficient than public key method and simpler to code )

* `remove_verifier(verifier_id)` - Returns the deposited amount and removes the verifier from the registry.

> * `get_verifier(verifier_id)` - Returns details for a given verifier.

Reviewer 8: ```suggestion
* `get_verifier(verifier_id)` - Returns details for a given verifier.
* `update_sources(code_hash, verifier_id, sources_json_url, signatures)` - Verifies that the signatures match the quorum detailed under **verifier registry** and forwards the message to **sources registry** with the url.
```

Notes: These are high level actions that will be translated to exact message TL/B's and getters later. Potential sharding of this contract to multiple contracts is an implementation detail.

## Sources registry contract

> A smart contract deployed to TON mainnet that holds a mapping between a **code hash + verifier id** to the **sources.json url** generated by this verifier.

Reviewer 8: ```suggestion
A contract that yields a source contract per **code hash + verifier id**, similar to nft collection or jetton minter.

The source contract stores the url of the sources.json file, and its address can be calculated client-side, without querying the blockchain.
```

### Actions

> * `update_sources(code_hash, verifier_id, sources_json_url, signatures)` - Verifies that the signatures match the quorum detailed under **verifier registry** and updates the **sources registry** with the url.

Reviewer 8: ```suggestion
* `update_sources(code_hash, verifier_id, sources_json_url, signatures)` - updates the **sources registry** with the url.
```

> - **TEP**: [91](https://github.com/ton-blockchain/TEPs/pull/91)

Reviewer 3: UQD73zOXk5RAywtZFUfAYfgGWjEd02snHr4Axfud90RdcbVe

## sources.json

JSON file provided by a specific **verifier** for a specific **contract code hash** containing the URLs of source-code files and verification attestations. Fields of this file include:

* `codeHash` - SHA256 of the code cell CellRepr (https://ton-blockchain.github.io/docs/tvm.pdf 3.1.5), e.g. E/XXoxbG124QU+iKxZtd5loHKjiEUTcdxcW+y7oT9Q4=
* `sources` - Array of sources files, each with the following fields:
    * `url` - Human-readable source-code URL on IPFS and later TON Storage, e.g. [ipfs://QmWQE1HhYuWieFcocfZZvK4gvJqgswNpj3zUCvdPNCC12T](https://tonsource.infura-ipfs.io/ipfs/QmWQE1HhYuWieFcocfZZvK4gvJqgswNpj3zUCvdPNCC12T)
    * `originalFilename` - e.g. "wallet-v3.fc" or "imports/stdlib.fc"
* `compilerType` - Compiler used to compile the verified sources, e.g. "func"
* `compilerVersion` - Version of compiler, e.g. "0.2.0"
* `commandLine` - Command-line used to compile sources, e.g. "func -o output.fif -SPA wallet-v3.fc stdlib.fc"
* `verificationTime` - Unix timestamp of when verification was made

> 

Reviewer 12: Why do we need to keep things so complicated? 

In my opinion, it is much more reasonable to use 1 link to the project in ipfs (onchain), and also to store hash of on-chain code giving it in get method together with verifier_id to make such contracts easier to index. Making a link to each source file is very difficult.

It is also necessary that this project contains a configuration file by the proposed standard [TEP: FunC project configuration file](https://github.com/ton-blockchain/TEPs/pull/83). I believe that this standard can unify all CLI tools and will work much more reliably for building projects.

## Verifier registry contract

> A smart contract deployed to TON mainnet that holds a mapping between a **verifier id** to the **verifier details** which include the list of backends, their public keys and quorum configuration. To prevent spam in this registry, we propose that each verifier will deposit in the contract a sum of 1,000-10,000 TON coin. This sum will be returned when the verifier unregisters.

Reviewer 8: An interesting point was whether to tie the staking amount to a [config param](https://github.com/ton-blockchain/ton/blob/35d17249e6b54d67a5781ebf26e4ee98e56c1e50/crypto/block/block.tlb#L648), in order to reflect changes in TON/USD price.

E.g. 10 * 1e9 * Gas prices (which amounts to 10K ton)

Author: It's a good idea, since the contract is immutable without any special admin role, this can normalize the deposit size in case TON USD price changes significantly

Reviewer 13: > An interesting point was whether to tie the staking amount to a [config param](https://github.com/ton-blockchain/ton/blob/35d17249e6b54d67a5781ebf26e4ee98e56c1e50/crypto/block/block.tlb#L648), in order to reflect changes in TON/USD price.
> 
> 
> 
> E.g. 10 * 1e9 * Gas price (currently 1,000) => 10K ton
> 
> 
> 
>

Reviewer 13: > > An interesting point was whether to tie the staking amount to a [config param](https://github.com/ton-blockchain/ton/blob/35d17249e6b54d67a5781ebf26e4ee98e56c1e50/crypto/block/block.tlb#L648), in order to reflect changes in TON/USD price.
> > 
> > 
> > 
> > E.g. 10 * 1e9 * Gas price (currently 1,000) => 10K ton
> > 
> > 
> > 
> > 
> 
>

Reviewer 13: > It's a good idea, since the contract is immutable without any special admin role, this can normalize the deposit size in case TON USD price changes significantly

Reviewer 13: > > It's a good idea, since the contract is immutable without any special admin role, this can normalize the deposit size in case TON USD price changes significantly
> 
>

Reviewer: Cool

#### Actions

* `update_verifier(verifier_id, backend_endpoints, quorum_config)` - If the verifier does not exist, ensures the required amount was deposited and adds to the registry. Otherwise updates details in the registry. The address that sends this update message is stored in the registry as the admin address for this verifier and only it can update. The quorum config contains the list of public keys and how many are needed for quorum.
* `remove_verifier(verifier_id)` - Returns the deposited amount and removes the verifier from the registry. Message can only be sent by the verifier admin.
* `get_verifier(verifier_id)` - Returns details for a given verifier.

Notes: These are high level actions that will be translated to exact message TL-B's and getters later. Potential sharding of this contract to multiple contracts is an implementation detail.

## Sources registry contract

A smart contract deployed to TON mainnet that holds a mapping between a **code hash + verifier id** to the **sources.json url** generated by this verifier.

#### Actions

> * `update_sources(code_hash, verifier_id, sources_json_url, signatures)` - Verifies that the signatures match the verifier's quorum detailed under **verifier registry** and updates the **sources registry** with the url.

Reviewer 8: The signed data should include a measure of preventing replay attack, such as a `valid_until` timestamp which can be verified as well.

Reviewer 14: ادم باش

Reviewer 15: UQBhbNvEfQE7V-CJrVKPGPOcg1IwGq984K5_yQhqdQkU7vJD
